gallery/visualtests/testUtils.js

```diff
@@ -177,6 +181,56 @@ module.exports = {
       await page.mouse.move(0, 0);
     }
 
+    async function fillFile(path, numBytes) {
+      const MAX_BUFFER_SIZE = 1 << 20, // 1 MiB
+          writeStream = fs.createWriteStream(path);
+
+      let buffer;
+
+      for (let i = 0; i < numBytes; i += MAX_BUFFER_SIZE) {
+        const bufferSize = Math.min(MAX_BUFFER_SIZE, numBytes - i);
+
+        if (!(buffer && buffer.length === bufferSize)) {
+          buffer = await Buffer.alloc(bufferSize);
+        }
+
+        writeStream.write(buffer);
+      }
+
+      return new Promise((resolve, reject) => {
+        writeStream.on('finish', resolve);
+        writeStream.on('error', reject);
+
+        writeStream.end();
+      });
+    }
+
+    let files, tmpDir;
+
+    async function buildUploadableFiles() {
```

Comment:
Because it affects the `files` and `tmpDir` variables that come from closure scope, this function has some gotchas.  Mainly, that it can only safely be called once before `cleanupFiles` is called.  If `buildUploadableFiles` were to be called twice before `cleanupFiles` is called, the directory from the first call would be forgotten and not cleaned up.  I think it would be better if these functions were not exported, and instead you export a `setupUploadableFiles` function which contains the `beforeAll` and `afterAll` calls that are currently in `nxMultiFileUpload.js`.  That function could return the `files` (or perhaps a getter function for the files) for use in the tests

Reply:
> Because it affects the `files` and `tmpDir` variables that come from closure scope, this function has some gotchas. Mainly, that it can only safely be called once before `cleanupFiles` is called. If `buildUploadableFiles` were to be called twice before `cleanupFiles` is called, the directory from the first call would be forgotten and not cleaned up. I think it would be better if these functions were not exported, and instead you export a `setupUploadableFiles` function which contains the `beforeAll` and `afterAll` calls that are currently in `nxMultiFileUpload.js`. That function could return the `files` (or perhaps a getter function for the files) for use in the tests

I'm trying to figure this out, and struggling a bit. So we would create and export a `setupUploadableFiles`, and inside have a `beforeAll` hook (calling `buildUploadableFiles`), an `afterAll` hook (calling `cleanupUploadableFiles`), and also return `files` (or a getter function created inside of `files`)  that are rendered _after_ the `beforeAll` is called so the function doesn't return an `undefined` value. I'm thinking I need to `await` the results of `setupUploadableFiles` inside of `nxMultiFileUpload.js`, but I read that `describe` blocks don't support async functions anymore, so not sure how to call this correctly. Do you have any suggestions of where I can begin and I can go from there?

Reply:
>  that are rendered after the beforeAll is called so the function doesn't return an undefined value.

The timing issue there is why I think it needs to return a getter function rather than the actual files.  That getter would only be intended for use within tests, and if the caller keeps to that rule the `files` should be defined by the time they call the getter.  The getter function itself can be created before the `files` are though.  It just has to close over the mutable variable in which the `files` will be stored once they're set up.

> I'm thinking I need to await the results of setupUploadableFiles

No it would need to be synchronous.  I didn't know `describe` ever supported `async`.

Reply:
> > that are rendered after the beforeAll is called so the function doesn't return an undefined value.
> 
> The timing issue there is why I think it needs to return a getter function rather than the actual files. That getter would only be intended for use within tests, and if the caller keeps to that rule the `files` should be defined by the time they call the getter. The getter function itself can be created before the `files` are though. It just has to close over the mutable variable in which the `files` will be stored once they're set up.
> 
Fixed here: [b8480bb](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/b8480bb8337e065be7273351326f3ad94f2affad), then separate `beforeEach` here: [3e7ba4d](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/3e7ba4dc210dfb0852aa016ab8ee509efc851823) The function to retrieve `files` is called in every `it` block.

Reply:
> then separate beforeEach here: [3e7ba4d](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/3e7ba4dc210dfb0852aa016ab8ee509efc851823) The function to retrieve files is called in every it block.

It seems like it would have been fine without that second commit.  But more importantly, that commit isn't right because for instance `focusTest(...)` doesn't test anything, it returns a function that, when invoked tests something.  So it would need to be `await focusTest(...)()`

Reply:
> It seems like it would have been fine without that second commit. But more importantly, that commit isn't right because for instance `focusTest(...)` doesn't test anything, it returns a function that, when invoked tests something. So it would need to be `await focusTest(...)()`.

Thanks for pointing that out, definitely missed it. I put `getFiles` back into a beforeEach here [7202d0c](https://github.com/sonatype/sonatype-react-shared-components/pull/850/commits/7202d0ccb270d5bbc02b63c74552054c85e41812)